File: src/main/java/ai/RuleSet.java
package ai;

import java.awt.Point;
import java.util.LinkedList;

import game.Click;
import game.ClickType;

public class RuleSet {

	private LinkedList<Rule> bases;
	
	public RuleSet(LinkedList<Rule> rules) {
		bases = new LinkedList<>();
		bases.addAll(rules);
		removeDuplicates();
		boolean change = false;
		do {
			change = false;
			for (int i = 0; i < bases.size(); i++) {
				for (int j = 0; j < bases.size(); j++) {
					if (i != j && bases.get(i).contains(bases.get(j))
							&& !bases.get(j).isEmpty()) {
						change = true;
						bases.get(i).subtractBy(bases.get(j));
					}
				}
			}
		} while (change);
		removeEmpty();
	}
	
	public LinkedList<Click> solve() {
		LinkedList<Click> clicks = new LinkedList<Click>();
		
		for (Rule r: bases) {
			if (r.getPoints().size() == r.getMines()) {
				for (Point p: r.getPoints()) {
					Click c = new Click(ClickType.FLAG, p);
					if (!clicks.contains(c)) {
						clicks.add(c);
					}
				}
			}
			if (r.getMines() == 0) {
				for (Point p: r.getPoints()) {
					Click c = new Click(ClickType.REVEAL, p);
					if (!clicks.contains(c)) {
						clicks.add(c);
					}
				}
			}
		}
		return clicks;
	}
	
	public void print() {
		LinkedList<Point> points = new LinkedList<>();
		for (Rule r: bases) {
			for (Point p: r.getPoints()) {
				if (!points.contains(p)) {
					points.add(p);
				}
			}
		}
		System.out.println(bases.size() + "x" + points.size());
		for (Rule r: bases) {
			for (Point p: points) {
				if (r.contains(p)) {
					System.out.print("X");
				}
				else {
					System.out.print("-");
				}
			}
			System.out.println(" " + r);
		}
		System.out.println();
	}

	public LinkedList<Rule> getRules() {
		return bases;
	}
	
	private void removeDuplicates() {
		for (int i = 0; i < bases.size(); i++) {
			for (int j = i + 1; j < bases.size(); j++) {
				if (bases.get(i).equals(bases.get(j))) {
					bases.remove(j);
					j--;
				}
			}
		}
	}
	
	private void removeEmpty() {
		for (int i = 0; i < bases.size(); i++) {
			if (bases.get(i).isEmpty()) {
				bases.remove(i);
				i--;
			}
		}
	}
}

File: src/main/kotlin/ai/RuleSet.kt
package ai

import game.Click
import game.ClickType

class RuleSet(
	inputRules: List<Rule>
) {

	val rules: MutableList<Rule> = inputRules.distinct().toMutableList()

	init {
		var change: Boolean
		do {
			change = false
			rules.indices.forEach { i ->
				(i + 1 until rules.size).forEach { j ->
					if (rules[j].points.isNotEmpty() && rules[i].contains(rules[j])) {
						change = true
						rules[i] -= rules[j]
					}
					if (rules[i].points.isNotEmpty() && rules[j].contains(rules[i])) {
						change = true
						rules[j] -= rules[i]
					}
				}
			}
			rules.removeIf { it.points.isEmpty() }
		} while (change)
	}

	fun solve(): Set<Click> {
		val clicks = mutableSetOf<Click>()
		rules.forEach { (points, mines) ->
			if (points.size == mines) {
				points.forEach { clicks.add(Click(ClickType.FLAG, it)) }
			}
			if (mines == 0) {
				points.forEach { clicks.add(Click(ClickType.REVEAL, it)) }
			}
		}
		return clicks
	}
}